gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/policy/PartitionAwareHiveRegistrationPolicy.java

```diff
@@ -0,0 +1,135 @@
+
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+
+import org.apache.gobblin.configuration.State;
+import org.apache.gobblin.hive.HivePartition;
+import org.apache.gobblin.hive.HiveRegistrationUnit;
+import org.apache.gobblin.hive.HiveTable;
+
+
+/**
+ * This is the partition aware implementation of {@link HiveRegistrationPolicyBase}.
+ * You can specify hive.partition.regex where the first match is the table location and the
+ * rest match are the partitions.
+ * The partition names can be specified with the hive.table.partition.keys property.
+ * The order of the partition names should match the order of regexp matches in the hive.partition.regexp expression.
+ *
+ * For example in the case of path: s3://testbucket/myawesomlogs/compacted/dt=20170101/hr=22/
+ * The hive.partition.regexp would look like: hive.partition.regex=(s3://testbucket/myawesomelogs/compacted/)dt=(.*)/hr=(.*)
+ * and the hive.table.partition.keys=dt,hr
+ *
+ * @author Tamas Nemeth
+ */
+
+public class PartitionAwareHiveRegistrationPolicy extends HiveRegistrationPolicyBase{
+  private static final Logger LOG = LoggerFactory.getLogger(PartitionAwareHiveRegistrationPolicy.class);
+
+  public static final String HIVE_PARTITION_REGEX = "hive.partition.regex";
+  public static final String HIVE_TABLE_PARTITION_KEYS = "hive.table.partition.keys";
+
+  public PartitionAwareHiveRegistrationPolicy(State props)
+      throws IOException {
+    super(props);
+  }
+
+  protected HiveTable getTable(Path path, String dbName, String tableName) throws IOException {
+    LOG.debug("Getting table definition for {}", tableName);
+
+    HiveTable table = super.getTable(path, dbName, tableName);
+
+    LOG.debug("Getting partition keys for {}", tableName);
+    if (!Strings.isNullOrEmpty(this.props.getProp(HIVE_TABLE_PARTITION_KEYS))) {
+      List<HiveRegistrationUnit.Column> partitionKeyColumns = Lists.<HiveRegistrationUnit.Column> newArrayList();
+
+      for (String key : this.props.getPropAsList(HIVE_TABLE_PARTITION_KEYS)) {
+        LOG.debug("Setting partition key {} for table {}", key, tableName);
+        partitionKeyColumns.add(new HiveRegistrationUnit.Column(key , "string", null));
+      }
+
+      table.setPartitionKeys(partitionKeyColumns);
+    }
+
+    return table;
+  }
+
+
+  protected Optional<HivePartition> getPartition(Path path, HiveTable table) throws IOException {
+
+    if (this.props.contains(HIVE_PARTITION_REGEX)) {
+      Pattern pattern = Pattern.compile(this.props.getProp(HIVE_PARTITION_REGEX));
+      List<String> partitionValues = Lists.newArrayList();
+
+      Matcher matcher = pattern.matcher(path.toString());
+      if (matcher.matches() && matcher.groupCount() >=2){
+        for (int i = 2; i <= matcher.groupCount(); i++) {
+          partitionValues.add(matcher.group(i));
+        }
```

Comment:
Hey Tamas, 
As we're specifying partition keys, shouldn't we iterate on them to make sure all the partition key values are filled. For ex.: 
say 
`partition.keys` = dt, hr, min
and the regex is = (some-path)/dt=*/hr=* (something like this). `
Although the table will be registered with different partition keys, we would pass lesser number of values to register a partition which would lead an exception from HiveMetastore's thrift API. However, if we do it the other way, then the HivePartitionBuilder would fail quite early as compared to the other fail scenario.

Reply:
Make sense I will fix this and fail early.